<a href="https://colab.research.google.com/github/indahkhairunnisah-afk/Tugas-2-Teknik-Pengambilan-Sampel-dan-Data-Wrangling/blob/main/Tugas_2_TPSDW_Data_Preparation.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Persiapan Dan Load Dataset

Langkah awal adalah memuat library yang dibutuhkan dan membaca dataset. Kita juga perlu mengubah karakter ? menjadi NaN agar library Pandas dapat mendeteksinya sebagai data yang hilang (missing value).


In [30]:
import pandas as pd
import numpy as np

In [31]:
# Memuat dataset
df = pd.read_csv('/content/data_tugas 2.csv')

mengubah karakter "?" menjadi NaN agar library Pandas dapat mendeteksinya sebagai data yang hilang (missing value).

In [32]:
# Mengganti karakter '?' dengan NaN agar terdeteksi sebagai missing value
df = df.replace('?', np.nan)

In [33]:
# Menampilkan 5 data teratas
df.head()

,normalized-losses,make,bahan-bakar,jumlah-pintu,body-style,wheel-bases,length,width,height
0,NaN,alfa-romero,gas,two,convertible,88.6,168.8,64.1,48.8
1,NaN,alfa-romero,gas,two,convertible,88.6,168.8,64.1,48.8
2,NaN,alfa-romero,gas,two,hatchback,94.5,171.2,65.5,52.4
3,164,audi,gas,four,sedan,99.8,176.6,66.2,54.3
4,164,audi,gas,four,sedan,99.4,176.6,66.4,54.3


In [34]:
#Menampilkan 5 data terbawah
df.tail()

,normalized-losses,make,bahan-bakar,jumlah-pintu,body-style,wheel-bases,length,width,height
200,95,volvo,gas,four,sedan,109.1,188.8,68.9,55.5
201,95,volvo,gas,four,sedan,109.1,188.8,68.8,55.5
202,95,volvo,gas,four,sedan,109.1,188.8,68.9,55.5
203,95,volvo,diesel,four,sedan,109.1,188.8,68.9,55.5
204,95,volvo,gas,four,sedan,109.1,188.8,68.9,55.5


#Menampilkan Jumlah Data Hilang

 Menghitung jumlah data yang kosong pada setiap kolom untuk mengetahui sebaran missing value.

 *Menghasilkan DataFrame boolean (True jika data kosong/NaN, False jika ada nilai).
 Menjumlahkan nilai True per kolom, sehingga diperoleh jumlah data hilang (missing value) di setiap variabel.*

In [35]:
# Menghitung jumlah missing value per variabel
missing_values = df.isnull().sum()

*Menampilkan tabel jumlah missing value untuk setiap variabel, sehingga dapat menampilkan distribusi data hilang di seluruh kolom.*

In [36]:
# Membuat tabel ringkasan
print("Tabel Jumlah Missing Value:")
print(missing_values)

Tabel Jumlah Missing Value:
normalized-losses    41
make                  4
bahan-bakar          13
jumlah-pintu         18
body-style           12
wheel-bases           3
length                8
width                 5
height                4
dtype: int64


*Mencari nama kolom dengan nilai maksimum (kolom yang paling banyak missing value). max() mengambil jumlah maksimum missing value tersebut.
Menampilkan Variabel mana yang paling banyak kehilangan data (missing Value), dan berapa jumlahnya.* *teks yang dimiringkan*

In [37]:
# Mencari variabel dengan missing value terbanyak
max_missing_col = missing_values.idxmax()
max_missing_val = missing_values.max()

print(f"\nInterpretasi: Variabel yang memiliki data hilang paling banyak adalah '{max_missing_col}' dengan jumlah {max_missing_val} data.")


Interpretasi: Variabel yang memiliki data hilang paling banyak adalah 'normalized-losses' dengan jumlah 41 data.


*Berdasarkan hasil pengecekan, terdapat beberapa variabel yang memiliki data hilang. Variabel normalized-losses memiliki jumlah data hilang paling banyak, yaitu sebanyak 41 data.*

#Mengganti Missing Value Pada Variabel Kategori

*Untuk variabel kategorik (make, bahan-bakar, jumlah-pintu, body-style), data yang hilang akan diganti dengan label baru yaitu "unknown".*

*Memilih variabel 'make', 'bahan-bakar', 'jumlah-pintu', 'body-style' dan mengganti nilai kosong (NaN) dengan unknown.*

In [38]:
# Daftar variabel kategori
cat_cols = ['make', 'bahan-bakar', 'jumlah-pintu', 'body-style']

# Mengganti NaN dengan "unknown"
for col in cat_cols:
    df[col] = df[col].fillna('unknown')

*Untuk setiap variabel kategori, ditampilkan jumlah kemunculan tiap kategori, dengan ini terlihat distribusi per kategori termasuk berapa banyak yang masuk kategori 'unknwon'*

In [39]:
print("Ringkasan Jumlah Kategori:")
for col in cat_cols:
    print(f"\nVariabel: {col}")
    print(df[col].value_counts())

Ringkasan Jumlah Kategori:

Variabel: make
make
toyota           32
mazda            17
nissan           17
honda            13
mitsubishi       12
volkswagen       12
volvo            11
subaru           11
peugot           11
dodge             9
mercedes-benz     8
bmw               8
plymouth          7
audi              7
saab              6
porsche           4
unknown           4
isuzu             4
alfa-romero       3
chevrolet         3
jaguar            3
renault           2
mercury           1
Name: count, dtype: int64

Variabel: bahan-bakar
bahan-bakar
gas        173
diesel      19
unknown     13
Name: count, dtype: int64

Variabel: jumlah-pintu
jumlah-pintu
four       101
two         86
unknown     18
Name: count, dtype: int64

Variabel: body-style
body-style
sedan          88
hatchback      67
wagon          24
unknown        12
hardtop         8
convertible     6
Name: count, dtype: int64


*Setelah proses pembersihan dilakukan, seluruh variabel kategori sudah bebas dari nilai kosong karena setiap NaN diganti dengan label "unknown". Misalnya, pada variabel bahan-bakar terdapat 13 data yang kini berlabel "unknown", dan pada jumlah-pintu ada 18 data dengan label serupa. Pendekatan ini dipilih agar semua baris data tetap dipertahankan tanpa harus melakukan imputasi yang bersifat subjektif, sehingga analisis tetap konsisten sekaligus transparan terhadap keberadaan data yang tidak lengkap.*

#Mengganti Missing Value pada Variabel Numerik

Untuk variabel numerik (length, width, height), kita akan melakukan imputasi menggunakan nilai rata-rata (mean).

In [40]:
# Daftar variabel numerik
num_cols = ['length', 'width', 'height']

# Mengubah tipe data ke float (untuk memastikan perhitungan mean akurat)
for col in num_cols:
    df[col] = pd.to_numeric(df[col], errors='coerce')

# Hitung rata-rata dan isi missing value
means = df[num_cols].mean()
df[num_cols] = df[num_cols].fillna(means)

print("Nilai rata-rata yang digunakan untuk imputasi:")
print(means)

Nilai rata-rata yang digunakan untuk imputasi:
length    173.902538
width      65.931000
height     53.737313
dtype: float64


*Nilai yang hilang pada data numerik telah diganti dengan rata-rata masing-masing kolom, sehingga variabel length diisi dengan 173.90, width dengan 65.93, dan height dengan 53.73. Pengisian menggunakan rata-rata ini bertujuan untuk menjaga konsistensi distribusi data, sehingga meskipun ada data yang sebelumnya kosong, pola keseluruhan tetap terjaga tanpa menimbulkan bias yang berarti.*

#Min-Max Normalization

 *Melakukan normalisasi pada variabel wheel-bases agar nilainya berada dalam rentang 0 hingga 1.*

*Setiap nilai x pada kolom wheel-bases dikurangi dengan nilai minimum kolom tersebut, Hasilnya kemudian dibagi dengan selisih antara nilai maksimum dan minimum, dengan ini semua nilai akan dipetakan ke rentang 0 sampai 1*

*Hasil normalisasi disimpan dalam kolom baru bernama wheel-bases-normalized, Kolom asli (wheel-bases) tetap ada, sehingga bisa dibandingkan dengan hasil normalisasi.*

In [41]:
# Rumus Min-Max: (x - min) / (max - min)
df['wheel-bases-normalized'] = (df['wheel-bases'] - df['wheel-bases'].min()) / (df['wheel-bases'].max() - df['wheel-bases'].min())

*Menampilkan 5 baris pertama dari kolom asli dan kolom hasil normalisasi, tujuannya agar menampilkan bisa melihat bagaimana nilai asli dipetakan ke skala 0-1.*

In [42]:
# Menampilkan perbandingan kolom asli dan hasil normalisasi
print(df[['wheel-bases', 'wheel-bases-normalized']].head())

   wheel-bases  wheel-bases-normalized
0         88.6                0.058309
1         88.6                0.058309
2         94.5                0.230321
3         99.8                0.384840
4         99.4                0.373178


Normalisasi Min-Max berhasil dilakukan pada kolom wheel-bases. Contohnya, nilai asli 88.6 kini telah bertransformasi menjadi sekitar 0.058 dalam skala 0-1. Normalisasi ini sangat berguna untuk algoritma machine learning agar fitur dengan skala besar tidak mendominasi fitur dengan skala kecil.

#Z-Score Normalization

*Melakukan standarisasi (Z-score normalization) pada variabel length, width, dan height sehingga memiliki rata-rata 0 dan standar deviasi 1.*

*Setiap nilai x dikurangi dengan rata-rata (mean) kolom, Hasilnya dibagi dengan standar deviasi (std) kolom tersebut,Nilai Z-score menunjukkan seberapa jauh suatu data dari rata-rata dalam satuan standar deviasi.*

*Dilakukan perulangan untuk tiga variabel: length, width, dan height, Hasil standarisasi disimpan dalam kolom baru dengan nama tambahan -zscore (misalnya length-zscore).*

In [43]:
# Rumus Z-score: (x - mean) / std
for col in ['length', 'width', 'height']:
    df[f'{col}-zscore'] = (df[col] - df[col].mean()) / df[col].std()


*Menampilkan 5 baris pertama dari nilai asli dan nilai hasil Z-score untuk dibandingkan.*

In [44]:
# Menampilkan hasil standarisasi
print(df[['length', 'length-zscore', 'width', 'width-zscore', 'height', 'height-zscore']].head())

   length  length-zscore  width  width-zscore  height  height-zscore
0   168.8      -0.428526   64.1     -0.867630    48.8      -2.038094
1   168.8      -0.428526   64.1     -0.867630    48.8      -2.038094
2   171.2      -0.226967   65.5     -0.204232    52.4      -0.552035
3   176.6       0.226541   66.2      0.127467    54.3       0.232274
4   176.6       0.226541   66.4      0.222238    54.3       0.232274


*Dengan standarisasi Z-score, data length, width, dan height kini berada dalam skala yang sama, yaitu rata-rata bernilai 0 dan standar deviasi bernilai 1. Jika nilainya negatif (seperti length-zscore -0.42), berarti nilai tersebut berada di bawah rata-rata populasi, sedangkan nilai positif berarti di atas rata-rata. Hal ini memudahkan analisis statistik maupun pemodelan machine learning, karena variabel dengan skala berbeda tidak lagi mendominasi perhitungan.*

*Menympan data frame df ke dalam sebuah file CSV bernama data_tugas_2_clean.csv, Dengan opsi index=False, indeks baris DataFrame tidak ikut ditulis ke file sehingga hanya kolom data yang tersimpan.*

In [45]:
df.to_csv('data_tugas_2_clean.csv', index=False)